File: examples/AbcEquivariantFilterExample.cpp
/**
 * @file AbcEquivariantFilterExample.cpp
 * @brief Demonstration of the Attitude-Bias-Calibration Equivariant Filter
 *
 * This demo shows the Equivariant Filter (EqF) for attitude estimation
 * with both gyroscope bias and sensor extrinsic calibration, based on the
 * paper: "Overcoming Bias: Equivariant Filter Design for Biased Attitude
 * Estimation with Online Calibration" by Fornasier et al.
 *
 * This example uses the simplified AbcEquivariantFilter class which
 * provides a clean interface for predict/update operations without requiring
 * manual computation of Jacobian matrices and innovation functions.
 *
 * @author Darshan Rajasekaran
 * @author Jennifer Oum
 * @author Rohan Bansal
 * @author Frank Dellaert
 * @date 2025
 */

#include <gtsam/slam/dataset.h>
#include <gtsam_unstable/geometry/ABC.h>
#include <gtsam_unstable/geometry/ABCEquivariantFilter.h>

#include <array>
#include <chrono>
#include <cmath>
#include <cstddef>
#include <exception>
#include <fstream>
#include <iostream>
#include <numeric>
#include <sstream>
#include <stdexcept>
#include <string>
#include <vector>

// Use namespace for convenience
using namespace gtsam;
constexpr size_t n = 1;  // Number of calibration states
using M = abc::State<n>;
using AbcFilter = abc::AbcEquivariantFilter<n>;

/// Measurement struct
struct Measurement {
  Unit3 y;           /// Measurement direction in sensor frame
  Unit3 d;           /// Known direction in global frame
  Matrix3 R;         /// Covariance matrix of the measurement
  int cal_idx = -1;  /// Calibration index (-1 for calibrated sensor)
};

/// Data structure for ground-truth, input and output data
struct Data {
  M xi;                     /// Ground-truth state
  Vector3 omega;            /// Angular velocity measurement
  Matrix6 inputCovariance;  /// Input noise covariance (6x6 matrix)
  std::vector<Measurement> measurements;  /// Output measurements
  int numMeasurements;                    /// Number of measurements
  double t;                               /// Time
  double dt;                              /// Time step
};

//========================================================================
// Data Processing Functions
//========================================================================

/**
 * Load data from CSV file into a vector of Data objects for the EqF
 *
 * CSV format:
 * - t: Time
 * - q_w, q_x, q_y, q_z: True attitude quaternion
 * - b_x, b_y, b_z: True bias
 * - cq_w_0, cq_x_0, cq_y_0, cq_z_0: True calibration quaternion
 * - w_x, w_y, w_z: Angular velocity measurements
 * - std_w_x, std_w_y, std_w_z: Angular velocity measurement standard deviations
 * - std_b_x, std_b_y, std_b_z: Bias process noise standard deviations
 * - y_x_0, y_y_0, y_z_0, y_x_1, y_y_1, y_z_1: Direction measurements
 * - std_y_x_0, std_y_y_0, std_y_z_0, std_y_x_1, std_y_y_1, std_y_z_1: Direction
 * measurement standard deviations
 * - d_x_0, d_y_0, d_z_0, d_x_1, d_y_1, d_z_1: Reference directions
 *
 */
std::vector<Data> loadDataFromCSV(const std::string& filename, int startRow = 0,
                                  int maxRows = -1, int downsample = 1);

/// Process data with EqF and print summary results
void processDataWithEqF(AbcFilter& filter, const std::vector<Data>& data_list,
                        int printInterval = 10);

//========================================================================
// Data Processing Functions Implementation
//========================================================================

/*
 * Loads the test data from the csv file
 * startRow First row to load based on csv, 0 by default
 * maxRows maximum rows to load, defaults to all rows
 * downsample Downsample factor, default 1
 * A list of data objects
 */

std::vector<Data> loadDataFromCSV(const std::string& filename, int startRow,
                                  int maxRows, int downsample) {
  std::vector<Data> data_list;
  std::ifstream file(filename);

  if (!file.is_open()) {
    throw std::runtime_error("Failed to open file: " + filename);
  }

  std::cout << "Loading data from " << filename << "..." << std::flush;

  std::string line;
  int lineNumber = 0;
  int rowCount = 0;
  int errorCount = 0;
  double prevTime = 0.0;

  // Skip header
  std::getline(file, line);
  lineNumber++;

  // Skip to startRow
  while (lineNumber < startRow && std::getline(file, line)) {
    lineNumber++;
  }

  // Read data
  while (std::getline(file, line) && (maxRows == -1 || rowCount < maxRows)) {
    lineNumber++;

    // Apply downsampling
    if ((lineNumber - startRow - 1) % downsample != 0) {
      continue;
    }

    std::istringstream ss(line);
    std::string token;
    std::vector<double> values;

    // Parse line into values
    while (std::getline(ss, token, ',')) {
      try {
        values.push_back(std::stod(token));
      } catch (const std::exception& e) {
        errorCount++;
        values.push_back(0.0);  // Use default value
      }
    }

    // Check if we have enough values
    if (values.size() < 39) {
      errorCount++;
      continue;
    }

    // Extract values
    double t = values[0];
    double dt = (rowCount == 0) ? 0.0 : t - prevTime;
    prevTime = t;

    // Create ground truth state
    Quaternion quat(values[1], values[2], values[3], values[4]);  // w, x, y, z
    Rot3 R = Rot3(quat);

    Vector3 b(values[5], values[6], values[7]);

    Quaternion calQuat(values[8], values[9], values[10],
                       values[11]);  // w, x, y, z
    std::array<Rot3, n> S = {Rot3(calQuat)};

    M xi(R, b, S);

    // Create input
    Vector3 omega(values[12], values[13], values[14]);

    // Create input covariance matrix (6x6)
    // First 3x3 block for angular velocity, second 3x3 block for bias process
    // noise
    Matrix6 inputCovariance = Matrix6::Zero();
    inputCovariance(0, 0) = values[15] * values[15];  // std_w_x^2
    inputCovariance(1, 1) = values[16] * values[16];  // std_w_y^2
    inputCovariance(2, 2) = values[17] * values[17];  // std_w_z^2
    inputCovariance(3, 3) = values[18] * values[18];  // std_b_x^2
    inputCovariance(4, 4) = values[19] * values[19];  // std_b_y^2
    inputCovariance(5, 5) = values[20] * values[20];  // std_b_z^2

    // Create measurements
    std::vector<Measurement> measurements;

    // First measurement (calibrated sensor, cal_idx = 0)
    Vector3 y0(values[21], values[22], values[23]);
    Vector3 d0(values[33], values[34], values[35]);

    // Normalize vectors if needed
    if (abs(y0.norm() - 1.0) > 1e-5) y0.normalize();
    if (abs(d0.norm() - 1.0) > 1e-5) d0.normalize();

    // Measurement covariance
    Matrix3 covY0 = Matrix3::Zero();
    covY0(0, 0) = values[27] * values[27];  // std_y_x_0^2
    covY0(1, 1) = values[28] * values[28];  // std_y_y_0^2
    covY0(2, 2) = values[29] * values[29];  // std_y_z_0^2

    // Create measurement
    measurements.push_back(Measurement{Unit3(y0), Unit3(d0), covY0, 0});

    // Second measurement (calibrated sensor, cal_idx = -1)
    Vector3 y1(values[24], values[25], values[26]);
    Vector3 d1(values[36], values[37], values[38]);

    // Normalize vectors if needed
    if (abs(y1.norm() - 1.0) > 1e-5) y1.normalize();
    if (abs(d1.norm() - 1.0) > 1e-5) d1.normalize();

    // Measurement covariance
    Matrix3 covY1 = Matrix3::Zero();
    covY1(0, 0) = values[30] * values[30];  // std_y_x_1^2
    covY1(1, 1) = values[31] * values[31];  // std_y_y_1^2
    covY1(2, 2) = values[32] * values[32];  // std_y_z_1^2

    // Create measurement
    measurements.push_back(Measurement{Unit3(y1), Unit3(d1), covY1, -1});

    // Create Data object and add to list
    data_list.push_back(
        Data{xi, omega, inputCovariance, measurements, 2, t, dt});

    rowCount++;

    // Show loading progress every 1000 rows
    if (rowCount % 1000 == 0) {
      std::cout << "." << std::flush;
    }
  }

  std::cout << " Done!" << std::endl;
  std::cout << "Loaded " << data_list.size() << " data points";

  if (errorCount > 0) {
    std::cout << " (" << errorCount << " errors encountered)";
  }

  std::cout << std::endl;

  return data_list;
}

/// Takes in the data and runs an EqF on it and reports the results
void processDataWithEqF(AbcFilter& filter, const std::vector<Data>& data_list,
                        int printInterval) {
  if (data_list.empty()) {
    std::cerr << "No data to process" << std::endl;
    return;
  }

  std::cout << "Processing " << data_list.size() << " data points with EqF..."
            << std::endl;

  // Track performance metrics
  std::vector<double> att_errors;
  std::vector<double> bias_errors;
  std::vector<double> cal_errors;

  // Track time for performance measurement
  auto start = std::chrono::high_resolution_clock::now();

  int totalMeasurements = 0;
  int validMeasurements = 0;

  // Define constant for converting radians to degrees
  const double RAD_TO_DEG = 180.0 / M_PI;

  // Print a progress indicator
  int progressStep = data_list.size() / 10;  // 10 progress updates
  if (progressStep < 1) progressStep = 1;

  std::cout << "Progress: ";

  for (size_t i = 0; i < data_list.size(); i++) {
    const Data& data = data_list[i];
    filter.predict(data.omega, data.inputCovariance, data.dt);

    // Process all measurements
    for (const auto& measurement : data.measurements) {
      totalMeasurements++;

      // Skip invalid measurements
      Vector3 y_vec = measurement.y.unitVector();
      Vector3 d_vec = measurement.d.unitVector();
      if (std::isnan(y_vec[0]) || std::isnan(y_vec[1]) ||
          std::isnan(y_vec[2]) || std::isnan(d_vec[0]) ||
          std::isnan(d_vec[1]) || std::isnan(d_vec[2])) {
        continue;
      }

      try {
        filter.update(measurement.y, measurement.d, measurement.R,
                      measurement.cal_idx);
        validMeasurements++;
      } catch (const std::exception& e) {
        std::cerr << "Error updating at t=" << data.t << ": " << e.what()
                  << std::endl;
      }
    }

    // Get current estimates using accessor methods
    Rot3 att_est = filter.attitude();
    Vector3 bias_est = filter.bias();
    Rot3 cal_est = filter.calibration(0);

    // Calculate errors
    Vector3 att_error = Rot3::Logmap(data.xi.R.between(att_est));
    Vector3 bias_error = bias_est - data.xi.b;
    Vector3 cal_error = Z_3x1;
    if (!data.xi.S.empty()) {
      cal_error = Rot3::Logmap(data.xi.S[0].between(cal_est));
    }

    // Store errors
    att_errors.push_back(att_error.norm());
    bias_errors.push_back(bias_error.norm());
    cal_errors.push_back(cal_error.norm());

    // Show progress dots
    if (i % progressStep == 0) {
      std::cout << "." << std::flush;
    }
  }

  std::cout << " Done!" << std::endl;

  auto end = std::chrono::high_resolution_clock::now();
  std::chrono::duration<double> elapsed = end - start;

  // Calculate average errors
  double avg_att_error = 0.0;
  double avg_bias_error = 0.0;
  double avg_cal_error = 0.0;

  if (!att_errors.empty()) {
    avg_att_error = std::accumulate(att_errors.begin(), att_errors.end(), 0.0) /
                    att_errors.size();
    avg_bias_error =
        std::accumulate(bias_errors.begin(), bias_errors.end(), 0.0) /
        bias_errors.size();
    avg_cal_error = std::accumulate(cal_errors.begin(), cal_errors.end(), 0.0) /
                    cal_errors.size();
  }

  // Calculate final errors from last data point
  const Data& final_data = data_list.back();
  Rot3 final_att_est = filter.attitude();
  Vector3 final_bias_est = filter.bias();
  Rot3 final_cal_est = filter.calibration(0);

  Vector3 final_att_error =
      Rot3::Logmap(final_data.xi.R.between(final_att_est));
  Vector3 final_bias_error = final_bias_est - final_data.xi.b;
  Vector3 final_cal_error = Z_3x1;
  if (!final_data.xi.S.empty()) {
    final_cal_error = Rot3::Logmap(final_data.xi.S[0].between(final_cal_est));
  }

  // Print summary statistics
  std::cout << "\n=== Filter Performance Summary ===" << std::endl;
  std::cout << "Processing time: " << elapsed.count() << " seconds"
            << std::endl;
  std::cout << "Processed measurements: " << totalMeasurements
            << " (valid: " << validMeasurements << ")" << std::endl;

  // Average errors
  std::cout << "\n-- Average Errors --" << std::endl;
  std::cout << "Attitude: " << (avg_att_error * RAD_TO_DEG) << "°" << std::endl;
  std::cout << "Bias: " << avg_bias_error << std::endl;
  std::cout << "Calibration: " << (avg_cal_error * RAD_TO_DEG) << "°"
            << std::endl;

  // Final errors
  std::cout << "\n-- Final Errors --" << std::endl;
  std::cout << "Attitude: " << (final_att_error.norm() * RAD_TO_DEG) << "°"
            << std::endl;
  std::cout << "Bias: " << final_bias_error.norm() << std::endl;
  std::cout << "Calibration: " << (final_cal_error.norm() * RAD_TO_DEG) << "°"
            << std::endl;

  // Print a brief comparison of final estimate vs ground truth
  std::cout << "\n-- Final State vs Ground Truth --" << std::endl;
  std::cout << "Attitude (RPY) - Estimate: "
            << (final_att_est.rpy() * RAD_TO_DEG).transpose()
            << "° | Truth: " << (final_data.xi.R.rpy() * RAD_TO_DEG).transpose()
            << "°" << std::endl;
  std::cout << "Bias - Estimate: " << final_bias_est.transpose()
            << " | Truth: " << final_data.xi.b.transpose() << std::endl;

  if (!final_data.xi.S.empty()) {
    std::cout << "Calibration (RPY) - Estimate: "
              << (final_cal_est.rpy() * RAD_TO_DEG).transpose() << "° | Truth: "
              << (final_data.xi.S[0].rpy() * RAD_TO_DEG).transpose() << "°"
              << std::endl;
  }
}

int main(int argc, char* argv[]) {
  std::cout << "ABC-EqF: Attitude-Bias-Calibration Equivariant Filter Demo"
            << std::endl;
  std::cout << "=============================================================="
            << std::endl;

  try {
    // Parse command line options
    std::string csvFilePath;
    int maxRows = -1;    // Process all rows by default
    int downsample = 1;  // No downsampling by default

    if (argc > 1) {
      csvFilePath = argv[1];
    } else {
      // Try to find the EQFdata file in the GTSAM examples directory
      try {
        csvFilePath = findExampleDataFile("EqFdata.csv");
      } catch (const std::exception& e) {
        std::cerr << "Error: Could not find EqFdata.csv" << std::endl;
        std::cerr << "Usage: " << argv[0]
                  << " [csv_file_path] [max_rows] [downsample]" << std::endl;
        return 1;
      }
    }

    // Optional command line parameters
    if (argc > 2) {
      maxRows = std::stoi(argv[2]);
    }

    if (argc > 3) {
      downsample = std::stoi(argv[3]);
    }

    // Load data from CSV file
    std::vector<Data> data =
        loadDataFromCSV(csvFilePath, 0, maxRows, downsample);

    if (data.empty()) {
      std::cerr << "No data available to process. Exiting." << std::endl;
      return 1;
    }

    // Initial covariance - larger values allow faster convergence
    Matrix initialSigma = Matrix::Identity(6 + 3 * n, 6 + 3 * n);
    initialSigma.diagonal().head<3>() =
        Vector3::Constant(0.1);  // Attitude uncertainty
    initialSigma.diagonal().segment<3>(3) =
        Vector3::Constant(0.01);  // Bias uncertainty
    initialSigma.diagonal().tail<3>() =
        Vector3::Constant(0.1);  // Calibration uncertainty

    // Create filter with initial covariance (starts at identity state)
    AbcFilter filter(initialSigma);

    // Process data
    processDataWithEqF(filter, data);

  } catch (const std::exception& e) {
    std::cerr << "Error: " << e.what() << std::endl;
    return 1;
  }

  std::cout << "\nEqF demonstration completed successfully." << std::endl;
  return 0;
}

File: python/gtsam/examples/AbcEquivariantFilterExample.py
"""
Python translation of examples/AbcEquivariantFilterExample.cpp.

Runs the Attitude-Bias-Calibration EqF demo using the wrapped C++
EquivariantFilter (ABC-specific wrapper).
"""

from __future__ import annotations

from dataclasses import dataclass
import csv
import math
from typing import List

import numpy as np
import gtsam
from gtsam import Rot3, Unit3
from gtsam.utils import findExampleDataFile


@dataclass
class MeasurementRecord:
    y: np.ndarray
    d: np.ndarray
    R: np.ndarray
    cal_idx: int


@dataclass
class DataRecord:
    R: Rot3
    b: np.ndarray
    cal_rot: Rot3
    omega: np.ndarray
    input_covariance: np.ndarray
    measurements: List[MeasurementRecord]
    t: float
    dt: float


def _normalize(v: np.ndarray) -> np.ndarray:
    n = np.linalg.norm(v)
    if n == 0.0:
        return v
    return v / n


def load_data_from_csv(
    filename: str, start_row: int = 0, max_rows: int = -1, downsample: int = 1
) -> List[DataRecord]:
    data_list: List[DataRecord] = []
    with open(filename, newline="") as csvfile:
        reader = csv.reader(csvfile)
        header = next(reader, None)
        if header is None:
            return data_list

        line_number = 1
        row_count = 0
        prev_time = 0.0

        for row in reader:
            line_number += 1
            if line_number < start_row:
                continue
            if ((line_number - start_row - 1) % downsample) != 0:
                continue
            if max_rows != -1 and row_count >= max_rows:
                break
            if len(row) < 39:
                continue

            values = [float(x) if x else 0.0 for x in row]

            t = values[0]
            dt = 0.0 if row_count == 0 else t - prev_time
            prev_time = t

            R = Rot3.Quaternion(values[1], values[2], values[3], values[4])
            b = np.array([values[5], values[6], values[7]])

            cal_rot = Rot3.Quaternion(values[8], values[9], values[10], values[11])

            omega = np.array([values[12], values[13], values[14]])

            input_covariance = np.zeros((6, 6))
            input_covariance[0, 0] = values[15] ** 2
            input_covariance[1, 1] = values[16] ** 2
            input_covariance[2, 2] = values[17] ** 2
            input_covariance[3, 3] = values[18] ** 2
            input_covariance[4, 4] = values[19] ** 2
            input_covariance[5, 5] = values[20] ** 2

            measurements: List[MeasurementRecord] = []

            y0 = _normalize(np.array([values[21], values[22], values[23]]))
            d0 = _normalize(np.array([values[33], values[34], values[35]]))
            cov_y0 = np.diag([values[27] ** 2, values[28] ** 2, values[29] ** 2])
            measurements.append(MeasurementRecord(y0, d0, cov_y0, 0))

            y1 = _normalize(np.array([values[24], values[25], values[26]]))
            d1 = _normalize(np.array([values[36], values[37], values[38]]))
            cov_y1 = np.diag([values[30] ** 2, values[31] ** 2, values[32] ** 2])
            measurements.append(MeasurementRecord(y1, d1, cov_y1, -1))

            data_list.append(
                DataRecord(
                    R=R,
                    b=b,
                    cal_rot=cal_rot,
                    omega=omega,
                    input_covariance=input_covariance,
                    measurements=measurements,
                    t=t,
                    dt=dt,
                )
            )
            row_count += 1

    return data_list


def process_data_with_eqf(
    filter_eqf: gtsam.abc.AbcEquivariantFilter1,
    data_list: List[DataRecord],
) -> None:
    if not data_list:
        print("No data to process")
        return

    print(f"Processing {len(data_list)} data points with EqF...")
    att_errors: List[float] = []
    bias_errors: List[float] = []
    cal_errors: List[float] = []

    total_measurements = 0
    valid_measurements = 0

    rad_to_deg = 180.0 / math.pi
    progress_step = max(len(data_list) // 10, 1)
    print("Progress: ", end="", flush=True)

    for i, data in enumerate(data_list):
        filter_eqf.predict(data.omega, data.input_covariance, data.dt)

        for measurement in data.measurements:
            total_measurements += 1
            if np.any(np.isnan(measurement.y)) or np.any(np.isnan(measurement.d)):
                continue
            try:
                y_unit = Unit3(measurement.y)
                d_unit = Unit3(measurement.d)
                filter_eqf.update(y_unit, d_unit, measurement.R, measurement.cal_idx)
                valid_measurements += 1
            except Exception:
                continue

        estimate_R = filter_eqf.attitude()
        estimate_b = np.array(filter_eqf.bias()).reshape(3)
        estimate_cal = filter_eqf.calibration(0)

        att_error = Rot3.Logmap(data.R.between(estimate_R))
        bias_error = estimate_b - data.b
        cal_error = np.zeros(3)
        cal_error = Rot3.Logmap(data.cal_rot.between(estimate_cal))

        att_errors.append(np.linalg.norm(att_error))
        bias_errors.append(np.linalg.norm(bias_error))
        cal_errors.append(np.linalg.norm(cal_error))

        if i % progress_step == 0:
            print(".", end="", flush=True)

    print(" Done!")

    avg_att_error = float(np.mean(att_errors)) if att_errors else 0.0
    avg_bias_error = float(np.mean(bias_errors)) if bias_errors else 0.0
    avg_cal_error = float(np.mean(cal_errors)) if cal_errors else 0.0

    final_data = data_list[-1]
    final_R = filter_eqf.attitude()
    final_b = np.array(filter_eqf.bias()).reshape(3)
    final_cal = filter_eqf.calibration(0)
    final_att_error = Rot3.Logmap(final_data.R.between(final_R))
    final_bias_error = final_b - final_data.b
    final_cal_error = Rot3.Logmap(final_data.cal_rot.between(final_cal))

    print("\n=== Filter Performance Summary ===")
    print(f"Processed measurements: {total_measurements} (valid: {valid_measurements})")

    print("\n-- Average Errors --")
    print(f"Attitude: {avg_att_error * rad_to_deg}°")
    print(f"Bias: {avg_bias_error}")
    print(f"Calibration: {avg_cal_error * rad_to_deg}°")

    print("\n-- Final Errors --")
    print(f"Attitude: {np.linalg.norm(final_att_error) * rad_to_deg}°")
    print(f"Bias: {np.linalg.norm(final_bias_error)}")
    print(f"Calibration: {np.linalg.norm(final_cal_error) * rad_to_deg}°")

    print("\n-- Final State vs Ground Truth --")
    print(
        "Attitude (RPY) - Estimate:",
        (final_R.rpy() * rad_to_deg),
        "° | Truth:",
        (final_data.R.rpy() * rad_to_deg),
        "°",
    )
    print("Bias - Estimate:", final_b, "| Truth:", final_data.b)
    print(
        "Calibration (RPY) - Estimate:",
        (final_cal.rpy() * rad_to_deg),
        "° | Truth:",
        (final_data.cal_rot.rpy() * rad_to_deg),
        "°",
    )


def main() -> None:
    print("ABC-EqF: Attitude-Bias-Calibration Equivariant Filter Demo")
    print("==============================================================")

    try:
        csv_file_path = findExampleDataFile("EqFdata.csv")
    except Exception:
        print("Error: Could not find EqFdata.csv")
        return

    data = load_data_from_csv(csv_file_path)
    if not data:
        print("No data available to process. Exiting.")
        return

    n_cal = 1
    m_sensors = 2
    initial_sigma = np.eye(6 + 3 * n_cal)
    initial_sigma[0:3, 0:3] = 0.1 * np.eye(3)
    initial_sigma[3:6, 3:6] = 0.01 * np.eye(3)
    initial_sigma[6:9, 6:9] = 0.1 * np.eye(3)

    filter_eqf = gtsam.abc.AbcEquivariantFilter1(initial_sigma)
    process_data_with_eqf(filter_eqf, data)

    print("\nEqF demonstration completed successfully.")


if __name__ == "__main__":
    main()


